lte/gateway/c/core/oai/tasks/mme_app/mme_app_bearer.c

```diff
@@ -905,10 +905,21 @@ void mme_app_handle_delete_session_rsp(
     ue_context_p->nb_active_pdn_contexts -= 1;
   }
 
+  pdn_cid_t pid =
+      ue_context_p->bearer_contexts[EBI_TO_INDEX(delete_sess_resp_pP->lbi)]
```

Comment:
This linked bearer ID is not in 3GPP GTPV2C DELETE_SESSION_RESPONSE message, so this will make the CUPS variant segv.

Reply:
Hi @lionelgo if linked bearer ID is not in 3GPP GTPV2C DELETE_SESSION_RESPONSE message, then how to find the pdn context for this delete session procedure?

Reply:
Hi @lionelgo  delete_sess_resp_pP->lbi was already being used, how it was not causing the issue for the CUPS variant?

Reply:
@lionelgo indeed `delete_sess_resp_pP->lbi` is used in several places in the mme code, are these code segments never executed by CUPS?

@rsarwad @lionelgo shall we simply sandwich these code segments between `#if EMBEDDED_SGW` and `#endif`?

Reply:
@rsarwad @lionelgo since this map is really used by the paging logic that is internal to Magma only and not used by OAI call flows, it should be safe to do what I suggest I think. @lionelgo could you please confirm?

Reply:
but I am also surprised that before these changes on line 909 within the if clause `(ue_context_p->emm_context.new_attach_info)`, the same lines of codes do exist.

Reply:
Hi @ulaskozat, delete_sess_resp_pP->lbi is also used for PDN disconnect scenario.https://github.com/magma/magma/blob/master/lte/gateway/c/core/oai/tasks/mme_app/mme_app_bearer.c#L934

Reply:
> @lionelgo indeed `delete_sess_resp_pP->lbi` is used in several places in the mme code, are these code segments never executed when CUPS is used?
> 
> @rsarwad @lionelgo shall we simply sandwich these code segments between `#if EMBEDDED_SGW` and `#endif`?

@ulaskozat Suggesting two ways
1.Moving this snippet of code as part removing PDN context; And writing to DB should be called on every message transaction as it done for UE context and shall write to DB only if content of map is modified. Currently we only write to DB while modifying to DB. 
 In case if we are not calling writing to DB on every message transaction. What's happening is; on mme restart; mme calls free_state() which calls mme_app_ue_context_free_content() which eventually calls mme_app_free_pdn_context(); their removes the ue_ip from map; since content of map is changed; it writes to DB with ue_ip removed. 
 2. We need to define two pdn removal functions; in one function_1(): Only in-memory contents are freed, called as part mme restart  and in function_2(), along with removal of in-memeory contexts, we shall write to DB also

Reply:
> Hi @lionelgo delete_sess_resp_pP->lbi was already being used, how it was not causing the issue for the CUPS variant?

Good question

Reply:
@lionelgo can we have a conclusion on this?

Reply:
@rsarwad I think point 1 introduces dependency between functions trying to maintain the state, if we can separate these logic into two functions (point 2) that is more clear for state changes, plus helps testability

Reply:
Hi @ulaskozat, Could you confirm for making these changes

Reply:
I am not sure if I understood the benefits of these options. Isn't the issue here that we use lbi and OAI code does not use it? cc @ardzoht @lionelgo @rsarwad

Reply:
In approach-2, mme shall remove ue_ip from ueip_imsi map, once mme deletes the PDN context.
 mme_app_free_pdn_context() is getting called when mme restarts, as part of clearing in-memeory contexts. In mme_app_remove_ue_ipv4_addr() ue_ip is removed from map and is also updated to DB. So mme_app_remove_ue_ipv4_addr() is called within mme_app_free_pdn_context(); ue_ip will not be written to DB. 
 
So thought of defining two functions to clear  PDN contexts; in function_1(): Only in-memory contents are freed, called as part mme restart and in function_2(), along with removal of in-memory contexts, we shall write to DB also.

Reply:
when mme is restarted, all in memory context is already gone and gets populated from Redis. I am not sure what clean up you are referring to here when mme restarts.

Reply:
@rsarwad point is different from lbi discussion, it is about freeing in-memory state on mme exit, @rsarwad can you specify when this situation happens `mme_app_remove_ue_ipv4_addr() ue_ip is removed from map and is also updated to DB`? It is not clear to me from your description

Reply:
On mme restart, mme_app_exit() is invoked, then it invokes clear_mme_nas_state(). Backtrace would something looks like, 
mme_app_exit()
clear_mme_nas_state()
 MmeNasStateManager::getInstance().free_state()
 clear_mme_nas_hashtables()
 
 While clearing hahtable entries; mme_app_state_free_ue_context() gets invoked which eventually call mme_app_free_pdn_context();

In mme_app_free_pdn_context(), mme_app_remove_ue_ipv4_addr() was called to remove ue_ip from map and was also updated to DB.
So added code to remove ue_ip from map on reception of Delete session response message; Based on lbi, ue_ip is fetched
To resolve the lbi dependency, I thought use different logic to remove ue_ip from and also update the DB.

Reply:
`MmeNasStateManager::getInstance().free_state()` does not need to free ueip_imsi_map as it is not a dynamically created data structure. Why we should worry about removing IP addresses in the map when process is exiting?

Reply:
map::clear() method can be used if/when needed.

Reply:
Yes we don't need to clear map contents as it doesn't have dynamic contexts. I shall call mme_app_remove_ue_ipv4_addr() within pdn_connectivity_delete(). So we don't have to rely on lbi within delete session response and on mme restart, map is not modified.